select() and select all() issue #944

DarkPrince304 opened the pull request:
select() and select all() are now able to accept two arguments solving #944. The second argument will denote the parent element inside which the search has to be performed. I haven't replaced `document.getElementById()` , `document.getElementsByClass()` ,`document.getElementsByTagName()`  for cases when only one argument is provided for performance and I have only used the `querySelector()`, and `querySelectorAll()` when two arguments are provided for `select()` and `selectAll()` respectively. Also the second argument works when it denotes the id of an element.

shiffman commented:
I will review this soon, thank you @DarkPrince304!  In the meantime, I'm seeing that there are merge conflicts, can you check?

Also, is there not a way to use `getElementById()` from somether other than `document`, i.e.

```javascript
 var elt = document.getElementById("container");
 var elts = elt.getElementsByTagName('p');
```

If I'm right about this I think this would be the preferred methodology.

DarkPrince304 commented:
@shiffman : Oh, my bad. I guess I misunderstood and implemented it using querySelector().

But no worries I have implemented it again using the method you suggested and its working. You can check it out in my latest commit :)

lmccart commented on lib/addons/p5.dom.js:
```diff
@@ -58,12 +58,23 @@
+      }else {
+        p = p.slice(1);
+        container = document.getElementById(p);
+      }
```
Just to simplify and remove redundancy, this could be rewritten:
```javascript
if (p[0] === 'string) {
  p = p.slice(1);
}
container = document.getElementById(p);

DarkPrince304 replied:
Shouldn't it be ` if (p[0] === '#' ) ` instead of ` if (p[0] === 'string') ` ?

lmccart replied:
also wondering, if they input a string without a # before, should we assume it's an id, or maybe it makes more sense to assume it's a tag name and just pick the first element returned by getElementsByTagName()?
I think this aligns more closely with the way the first argument to this function works.

lmccart replied:
yes sorry thats what i meant

DarkPrince304 replied:
I think we should treat it as an id. Since there can be more than one same type of tags and that would mean searching for element inside two parent elements making things complex.

lmccart replied:
yes but with the first argument, if there is no # it is assumed to be a tag. so we could ignore completely if there's no # but it feels weird to have two different meanings for a plain string as arguments to the same function.

DarkPrince304 replied:
You are right about that. What should we do about it for now?

lmccart commented on lib/addons/p5.dom.js:
```diff
@@ -58,12 +58,23 @@
+        p = p.slice(1);
+        container = document.getElementById(p);
+      }
+    }else if (typeof p === 'object'){
```
I would change this to:
```javascript
} else if (`p instanceof 'p5.Element') {
```
to check that it's not some other random object.

lmccart commented on lib/addons/p5.dom.js:
```diff
@@ -58,12 +58,23 @@
+      }
+    }else if (typeof p === 'object'){
+      container = document.getElementById(p.elt.id);
+    }
```
I think we should support one more case, which is passing in an html dom element direct from the page. that is, something obtained manually via getElementById:
```javascript
var elt = document.getElementById('container');
select("#blah", elt);
```

the check for this case would be:
```javascript
} else if (p instanceof HTMLElement) {
  container = p;
}

lmccart replied:
This is actually not necessary, since we already have access to the element, so we can simplify line 73 to be:
```javascript
container = p.elt;
```

lmccart commented on lib/addons/p5.dom.js:
```diff
@@ -58,12 +58,23 @@
+    var container = document;
+    if (typeof p === 'string'){
+      if(p[0] !== '#'){
+        container = document.getElementById(p);
```
we should probably check that the result is not null. if it is, we should keep container = document, rather than setting it equal to the result of the search.

lmccart commented on lib/addons/p5.dom.js:
```diff
@@ -112,15 +123,26 @@
     var res;
     var str;
+    var container = document;
+    if (typeof p === 'string'){
```
same comments as above apply here, but seeing this it makes me realize we should probably just have one internal helper function (something like `getParentElement()`) that can be called from both select() and selectAll().

DarkPrince304 commented:
@lmccart I have made the desired changes. Please let me know if I am missing something :)

lmccart commented:
looks great, thanks!

mixemer commented:
Can we use select() and selectAll methods just like querySelector()? For example, I tried this and it did not work: const audio = select('audio[data-key="'+keyCode+'"]')
